sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/ReservationDiscipline.java

```diff
@@ -0,0 +1,127 @@
+package com.mesosphere.sdk.scheduler.multi;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+
+import com.mesosphere.sdk.offer.LoggingUtils;
+import com.mesosphere.sdk.scheduler.AbstractScheduler;
+import com.mesosphere.sdk.scheduler.SchedulerConfig;
+import com.mesosphere.sdk.scheduler.MesosEventClient.ClientStatusResponse;
+import com.mesosphere.sdk.storage.Persister;
+import com.mesosphere.sdk.storage.PersisterException;
+
+/**
+ * Offer discipline implementation which only allows a subset of N deploying services to receive offers at the same
+ * time. This avoids a potential deadlock where two deploying services in the same scheduler both want to consume the
+ * same resource. If the limit is set to 1, then only one of those services would be growing footprint at a time.
+ *
+ * The limit value itself is a tradeoff between safety, speed, cluster topology, and any quotas configured by the
+ * operator. The "safest" value is 1, but in practice it may be perfectly safe to use a greater value.
+ */
+public class ReservationDiscipline implements OfferDiscipline {
+
+    private static final Logger LOGGER = LoggingUtils.getLogger(ReservationDiscipline.class);
+
+    /**
+     * A limit on the number of reserving services, or <=0 for no limit.
+     */
+    private final int reservingMax;
+
+    /**
+     * Storage of which services have a RESERVING status at a time, or an empty optional to disable (no limit). We may
+     * limit the number of services that can be RESERVING at the same time. We persist this information to avoid
+     * thrashing across scheduler restarts.
+     */
+    private final DisciplineSelectionStore selectionStore;
+
+    private Optional<Set<String>> selectedReservingServices;
+
+    /**
+     * Use {@link OfferDiscipline#create(SchedulerConfig, Persister)} to create a new instance.
+     *
+     * @param reserveLimit the number of services which may consume footprint simultaneously
+     * @param persister the persister where the list of allowed services will be persisted across restarts
+     * @throws PersisterException if initialization with the persister failed
+     */
+    ReservationDiscipline(int reserveLimit, DisciplineSelectionStore selectionStore) {
+        if (reserveLimit <= 0) {
+            throw new IllegalArgumentException(
+                    "INTERNAL ERROR: Reservation limit must be 1 or greater, was: " + reserveLimit);
+        }
+        this.reservingMax = reserveLimit;
+        this.selectionStore = selectionStore;
+        this.selectedReservingServices = Optional.empty();
+    }
+
+    /**
+     * Prunes any entries from the persister for services which no longer exist.
+     */
+    @Override
+    public void updateServices(Collection<AbstractScheduler> services) throws PersisterException {
```

Comment:
What happens if the MultiService is running with `reservingMax == 10`, 10 services are persisted as reserving, the MultiService is restarted and `reservingMax == 5`?

I think we need to truncate the stored list in that case, no?